conu/backend/nspawn/image.py

```diff
@@ -0,0 +1,332 @@
+        if self.pull_policy == ImagePullPolicy.ALWAYS:
+            logger.debug("pull policy set to 'always', pulling the image")
+            self.pull()
+        elif self.pull_policy == ImagePullPolicy.IF_NOT_PRESENT and not self.is_present():
+            logger.debug("pull policy set to 'if_not_present' and image is not present, "
+                         "pulling the image")
+            self.pull()
+        elif self.pull_policy == ImagePullPolicy.NEVER:
+            logger.debug("pull policy set to 'never'")
+
+    def __repr__(self):
+        return "DockerImage(repository=%s, tag=%s)" % (self.name, self.tag)
+
+    def __str__(self):
+        return self.get_full_name()
+
+    def get_full_name(self):
+        """
+        Provide full, complete image name
+
+        :return: str
+        """
+        return " ".join([self.name, self.tag])
+
+    def _is_local(self):
+        return os.path.exists(self.name)
+
+    def get_id(self):
+        """
+        get unique identifier of this image
+
+        :return: str
+        """
+
+        if self._id is None:
+            tags = ""
+            name = self.name.encode("ascii")
+            if self._is_local():
+                basename = os.path.basename(name)
+                if basename.endswith(".raw"):
+                    basename = basename[0:-4]
+                if self.special_separator in basename:
+                    name, tags = basename.split(self.special_separator, 1)
+                else:
+                    name = basename
+            if tags:
+                self.tag = self.special_separator.join([tags, self.tag])
+            self._id = "{PREFIX}{HASH}{SEP}{TAGS}".format(PREFIX=constants.CONU_ARTIFACT_TAG,
+                                                          HASH=hashlib.sha512(name).hexdigest()[:10],
+                                                          SEP=self.special_separator,
+                                                          TAGS=self.tag)
+        return self._id
+
+    def is_present(self):
+        return self.get_id() in NspawnImage.list_all()
+
+    def pull(self):
+        """
+        Pull this image from URL. Raises an exception if the image is not found in
+        the registry.
+
+        :return: None
+        """
+        ident = self.get_id()
+        try:
+            if self._is_local():
+                logger.debug("Try to pull local file: {} -> {}".format(self.name, ident))
+                run_cmd(["machinectl", "--verify=no", "import-raw", self.name, ident])
+            else:
+                logger.debug("Try to pull URL: {} -> {}".format(self.name, ident))
+                print (" ".join(["machinectl", "--verify=no",  "pull-raw", self.name, ident]))
+                run_cmd(["machinectl", "--verify=no",  "pull-raw", self.name, ident])
+        except ValueError as error:
+            raise ConuException("There was an error while pulling the image %s: %s", self.name, error)
+
+    @staticmethod
+    def list_all():
```

Comment:
another great candidate for a generic function

Reply:
sure, it is mentioned in TODO: